core/src/main/java/apoc/nodes/Nodes.java

```diff
@@ -67,6 +75,58 @@
 
     @Context
     public Pools pools;
+    
+    @Procedure
+    @Description("CALL apoc.nodes.cycles([nodes], 'type', $config) - Detect all path cycles from node list")
+    public Stream<PathResult> cycles(@Name("nodes") List<Node> nodes, @Name(value = "config",defaultValue = "{}") Map<String, Object> config) {
+        NodesConfig conf = new NodesConfig(config);
+        final List<String> types = conf.getRelTypes();
+        Stream<Path> paths = nodes.stream().flatMap(start -> {
+            boolean allRels = types.isEmpty();
+            final RelationshipType[] relTypes = types.stream().map(RelationshipType::withName).toArray(RelationshipType[]::new);
+            final Iterable<Relationship> relationships = allRels
+                    ? start.getRelationships(Direction.OUTGOING)
+                    : start.getRelationships(Direction.OUTGOING, relTypes);
+
+            PathExpanderBuilder expanderBuilder;
+            if (allRels) {
+                expanderBuilder = PathExpanderBuilder.allTypes(Direction.OUTGOING);
+            } else {
+                expanderBuilder = PathExpanderBuilder.empty();
+                for (RelationshipType relType: relTypes) {
+                    expanderBuilder = expanderBuilder.add(relType, Direction.OUTGOING);
+                }
+            }
+            final PathExpander<Path> pathExpander = expanderBuilder.build();
+
+            PathFinder<Path> finder = GraphAlgoFactory.shortestPath(
+                    new BasicEvaluationContext(tx, db),
+                    pathExpander,
+                    conf.getMaxDepth());
+            Map<Long, List<Long>> dups = new HashMap<>();
+            return Iterables.stream(relationships)
+                    // to prevent duplicated (start and end nodes with double-rels)
+                    .filter(relationship -> {
+                        final List<Long> nodeDups = dups.computeIfAbsent(relationship.getStartNodeId(), (key) -> new ArrayList<>());
+                        if (nodeDups.contains(relationship.getEndNodeId())) {
+                            return false;
+                        }
+                        nodeDups.add(relationship.getEndNodeId());
+                        return true;
+                    })
+                    .flatMap(relationship -> {
+                        final Path path = finder.findSinglePath(relationship.getEndNode(), start);
+                        if (path == null) return Stream.empty();
+                        PathImpl.Builder builder = new PathImpl.Builder(start)
```

Comment:
A good opportunity to use `VirtualPath` here (maybe time to introduce a `Builder`)